src/plugins/liveobjects/livecounter.ts

```diff
@@ -90,19 +87,72 @@ export class LiveCounter extends LiveObject<LiveCounterData, LiveCounterUpdate>
         );
     }
 
-    this.setRegionalTimeserial(opRegionalTimeserial);
     this.notifyUpdated(update);
   }
 
+  /**
+   * @internal
+   */
+  overrideWithStateObject(stateObject: StateObject): LiveCounterUpdate {
+    if (stateObject.objectId !== this.getObjectId()) {
+      throw new this._client.ErrorInfo(
+        `Invalid state object: state object objectId=${stateObject.objectId}; LiveCounter objectId=${this.getObjectId()}`,
+        50000,
+        500,
+      );
+    }
+
+    if (!this._client.Utils.isNil(stateObject.createOp)) {
+      // it is expected that create operation can be missing in the state object, so only validate it when it exists
+      if (stateObject.createOp.objectId !== this.getObjectId()) {
+        throw new this._client.ErrorInfo(
+          `Invalid state object: state object createOp objectId=${stateObject.createOp?.objectId}; LiveCounter objectId=${this.getObjectId()}`,
+          50000,
+          500,
+        );
+      }
+
+      if (stateObject.createOp.action !== StateOperationAction.COUNTER_CREATE) {
+        throw new this._client.ErrorInfo(
+          `Invalid state object: state object createOp action=${stateObject.createOp?.action}; LiveCounter objectId=${this.getObjectId()}`,
+          50000,
+          500,
+        );
+      }
+    }
+
+    const previousDataRef = this._dataRef;
+    // override all relevant data for this object with data from the state object
+    this._createOperationIsMerged = false;
+    this._dataRef = { data: stateObject.counter?.count ?? 0 };
+    this._siteTimeserials = this._timeserialMapFromStringMap(stateObject.siteTimeserials);
+    if (!this._client.Utils.isNil(stateObject.createOp)) {
+      this._mergeInitialDataFromCreateOperation(stateObject.createOp);
+    }
+
+    return this._updateFromDataDiff(previousDataRef, this._dataRef);
+  }
+
   protected _getZeroValueData(): LiveCounterData {
     return { data: 0 };
   }
 
-  protected _updateFromDataDiff(currentDataRef: LiveCounterData, newDataRef: LiveCounterData): LiveCounterUpdate {
-    const counterDiff = newDataRef.data - currentDataRef.data;
+  protected _updateFromDataDiff(prevDataRef: LiveCounterData, newDataRef: LiveCounterData): LiveCounterUpdate {
+    const counterDiff = newDataRef.data - prevDataRef.data;
     return { update: { inc: counterDiff } };
   }
 
+  protected _mergeInitialDataFromCreateOperation(stateOperation: StateOperation): LiveCounterUpdate {
+    // if a counter object is missing for the COUNTER_CREATE op, the initial value is implicitly 0 in this case.
+    // note that it is intentional to SUM the incoming count from the create op.
+    // if we got here, it means that current counter instance is missing the initial value in its data reference,
+    // which we're going to add now.
+    this._dataRef.data += stateOperation.counter?.count ?? 0;
+    this._createOperationIsMerged = true;
```

Comment:
this is the merging part I've mentionted in the PR description (same approach for the LiveMap).
I couldn't think of a situation where this might bite us on the SDK side, so merging it once and setting a flag for it seems most reasonable to save on merges on each read.
wdyt @mschristensen ?

Reply:
Yes I think this works :)